Revert managers and operators in tearDown method

shilpakodli opened the pull request:
#### Describe Bug or Feature
Currently PTL is not reverting server attributes "managers" and "operators" at the end of the test case. 


#### Describe Your Change
Unsetting server attributes "managers" and "operators" in cleanup_jobs method which is called in tearDown function.


#### Attach Test Logs or Output
[pbs_managers_operators_execution_logs.txt](https://github.com/PBSPro/pbspro/files/3043326/pbs_managers_operators_execution_logs.txt)


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8328,6 +8328,11 @@ def cleanup_jobs(self, extend=None, runas=None):
+        revert_server_attribs = ['managers', 'operators']
+
+        for attrib in revert_server_attribs:
+            self.manager(MGR_CMD_UNSET, SERVER, attrib, sudo=True)
```
There is no need for the for loop.  You can pass a attrib list to manager().

hirenvadalia replied:
Why we are reverting server attribute in cleanup_jobs()? This method is to clean up jobs only, right.

shilpakodli replied:
Since there is code relating setting server's attribute "managers" to current user, I have made changes at same place.

bhroam replied:
@latha-subramanian had to revert the managers here so the jobs could be deleted.

hirenvadalia replied:
@latha-subramanian Can you tell us why you had to revert the managers here? As far as I know in Server.revert_to_defaults() cleanup jobs happen first then it reverts server attributes. So why we need to revert the managers here (I mean before reverting server attributes)?

latha-subramanian replied:
@hirenvadalia ,Sorry for missing this earlier. cleanup_jobs() wants mangers attribute to be set in order to delete jobs.  Some tests unset managers attribute(to test some specific scenarios) . Hence cleanup_jobs() had to set it back.  Did it answer your question?

hirenvadalia replied:
@latha-subramanian If I understand your answer correct then we need manager attribute set to current user to delete other user's job in PBS right? if that's the only case then I don't think we need set manager here because Server.manager() can access sudo and/or runas argument and in PBS root user can delete any user's job.

latha-subramanian replied:
We would also want to kill all the job processes (in _cleanup_large_num_jobs()). If I remember right, I wasn't able to do this without manager attribute set.

hirenvadalia replied:
@latha-subramanian Even in case of that you don't need to set manager attribute because _cleanup_large_num_jobs() has runas argument and we can pass root in that and all 3 API (Server.status, DshUtils.run_cmd and Server.delete) which are called by _cleanup_large_num_jobs can take runas argument. So I still don't think we need to set manager attribute to delete jobs before we revert server attribute.

latha-subramanian replied:
@hirenvadalia , Yes. I agree for the status(),run_cmd() and delete functions 'runas' argument will suffice. But there is also a qmgr call to make scheduling 'false' and 'true' in cleanup_jobs().  Is there a way out for this without setting the 'manager' attribute ?

latha-subramanian replied:
@hirenvadalia , I checked manager() call was successful with 'runas' as root. I will raise a bug and get this one fixed. Thanks again for pointing this out !

bhroam commented on test/tests/selftest/pbs_managers_operators_ext.py:
```diff
@@ -60,3 +60,28 @@ def test_managers_unset_setup(self):
+        TestSelf.tearDown(self)
+        self.server.expect(SERVER, attrib=attr)
+        try:
+            self.server.expect(SERVER, attrib=a, max_attempts=5)
```
I don't think the try/except is needed.  I think you can do an expect() with op=NE with the same effect.

shilpakodli commented:
@bhroam and @hirenvadalia I have addressed your review comments. Please find attached latest execution logs.
[pbs_managers_operators_execution_logs.txt](https://github.com/PBSPro/pbspro/files/3053283/pbs_managers_operators_execution_logs.txt)

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11163,6 +11158,20 @@ def revert_to_defaults(self):
         self.resource_group = None
         return self.isUp()
 
+    def revert_server_attribs(self):
```
I dislike this name since it could be confused with revert_to_defaults().  How about something more specific: revert_server_tearDown() or something like that.

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11163,6 +11158,20 @@ def revert_to_defaults(self):
+        self.manager(MGR_CMD_UNSET, SERVER, unset_list, sudo=True)
+
+        current_user = pwd.getpwuid(os.getuid())[0]
+        a = {ATTR_managers: (INCR, current_user + '@*')}
```
I don't know if it is worthwhile, but if the user is already in managers, this spits out an error.  It's not a harmful error though.  We want the user in managers.  We could get rid of this error by checking first, but it will require a stat call to the server.  Is it worth it to avoid printing an error?

shilpakodli replied:
@bhroam I am unsetting managers attribute before I set it to current user. I am doing this because it might be set to some other users in addition to current user as part of few test cases. I am also handling error which will be likely thrown from "self.manager(MGR_CMD_SET, SERVER, a, sudo=True)"

bhroam replied:
Great, thanks

shilpakodli commented:
@bhroam I have addressed your review comments. Please let me know if ant changes needs to be done.

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5190,6 +5190,20 @@ def revert_to_defaults(self, reverthooks=True, revertqueues=True,
                 self.manager(MGR_CMD_DELETE, RSC, id=rescs)
         return True
 
+    def revert_server_attribs(self):
```
I don't see the name being changed to revert_server_teardown

shilpakodli replied:
Sorry Bhroam, I missed to push one file change. Please review it now.

shilpakodli commented:
Closing this PR as this is already covered in setUp.
